# Sessió 1: Python I

Indicacions:
+ Cada exercici s'ha de poder respondre executant una única funció (tot i que aquesta funció pot cridar altres funcions si és necessari).

## Nombres binaris fins a $N$

Donat un número $N$, creat tots els nombres binaris entre 1 i  $N$, ambdós inclosos. 

Pots fer servir la funció ``bin``. La funció ``bin(x)`` transforma $x$ en la seva forma binària. Observa els següents exemples:

In [3]:
a = 5
ab = bin(a)
print (a, ab)
print(bin(5+3))
type(bin(5+3))

5 0b101
0b1000


str

**Pregunta 1**: Escriu una funció que imprimeixi la representació binària de tots els nombres enters entre 1 i $N$. Quina solució hauria de donar per als següents valors: 
        
- $N$ = 5
- $N$ = 10

Intenta fer una versió del programa que no mostri la capçalera ``0b`` i que ompli de zeros el principi de cada nombre si no ocupem tots els bits (és a dir, que imprimeixi ``0001`` per l'enter 1).


In [3]:
def generaBinari(N):
    """
    Aquesta funció genera els nombres binaris
    entre 1 i N
    :param N el valor més gran pels binaris
    :return mostra per pantalla els nombres binari

    """
    for i in range (1,N+1):            
        print (bin(i)[2:].zfill(4))      
        

    

In [4]:
generaBinari (10)

0001
0010
0011
0100
0101
0110
0111
1000
1001
1010


N=5 01 10 11 100 101
N= 01 10 11 100 101 110 111 1000 1001 1010

**Pregunta 2**: Per a $N$ = 3 quantes operacions es fan? indica les operacions que es fan a cada línia, i quantes vegades es repeteix aquella línia.

*Escriu la teva resposta aquí*: Per a n=3, es repeteix el bucle 3 cops.

## El 8 és primer o és darrer?

Donada una llista de nombres enters, retornar ``True`` si el nombre 8 és el primer element de la llista o és el darrer. Retornar ``False`` altrament.

**Pregunta 1**: Posa 4 exemples de llistes, dues que retornin ``True`` i dues que retornin ``False``.

In [23]:
llista1=[8,1,2,3] #true
llista2=[1,2,3,8] #true
llista3=[2,3,4,5] # false
llista4=[10,7,4] #false

print (llista1[0]==8 or llista1[-1]==8)
print (llista2[0]==8 or llista2[-1]==8)
print (llista3[0]==8 or llista3[-1]==8)
print (llista4[0]==8 or llista4[-1]==8)




True
True
False
False


In [7]:
def primerDarrer8(llista):
    """
    Aquesta funció, donada una llista d’enters,
    retorna True si 8 és el primer o el darrer element de la llista, 
    retorna False altrament
    :param llista una llista d'enters
    :return: True si 8 és a l'inici o final, False altrament
    """
    if (llista[0]==8 or llista[-1]==8):
        return True
    return False


In [8]:
# Test de la funció
primerDarrer8([8,1,2,3])

# Hauria de retornar: True

True

## A mig camí

Donades dues llistes d'enters, cadascuna de longitud 3, retorna una nova llista que contingui els elements del mig de cada llista.

Per exemple:
- [4, 5, 6], [7, 8, 9] => [5, 8]
- [6, 6, 6], [1, 2, 3] => [6, 2]

**Pregunta 1**: Inventa't 3 parelles de llistes que retornin:
    
- [3, 3]
- [-1, 2]
- [0, 123]

In [ ]:
llista1=[1,3,2] 
llista2=[2,3,1]

llista3=[4,-1,5] 
llista4=[7,2,9]

llista5=[5,0,5] 
llista6=[100,123,9]



In [38]:
def aMigCami(llista1,llista2):
    """
    Aquesta funció, donades dues llistes d'enters,
    retorna una nova llista formada pels seus dos elements del mig
    :param llista1 una llista d'enters
    :param llista2 una llista d'enters
    :return: una llista formada pels dos elements del mig de llista1 i llista2
    """
    llista3=[]
    llista3.append(llista1[1])
    llista3.append(llista2[1])
    

    return llista3






    

In [39]:
# Test de la funció
aMigCami([10,11,12],[6,5,4])

# Hauria de retornar: [11,5]

[11, 5]

## Prou cerveses?

Quan els estudiants monten una festa els agrada tenir cerveses. Per a què una festa tingui èxit cal que hi hagi entre 50 i 100 cerveses, excepte en el cap de setmana en el que no hi ha un límit superior en el nombre de cerveses.

Escriu una funció '''prouCerveses''' en la que donat un nombre de cerveses i un booleà indicant si és cap de setmana o no, retorni si la festa serà un èxit.

for i in range (0, len(llista))
llista [i]
for j in range (i+1, len(llista))       apuntat a classe 

**Pregunta 1**: Proposa uns valors donats a la funció prouCerveses que provoquin un èxit de festa entre setmana i un èxit de festa el cap de setmana.

In [37]:
def prouCerveses(numCerveses,esCapDeSetmana):
    """
    Aquesta funció, donada una quantitat de cerveses,
    i un booleà que indica que és cap de setmana (True = cap de setmana)
    retorna si la festa serà un èxit.
    La festa serà un èxit si entre setmana hi ha entre 50 i 100 cerveses, 
    o si el cap de setmana hi ha més de 50 cerveses.
    :param numCerveses enter que indica el nombre de cerveses
    :param esCapDeSetmana booleà que indica si és cap de setmana
    :return: True si el nombre de cerveses està en els límits indicats
         False altrament
    """

    for i in range(50,100):
        if numCerveses<50:
            return False
        elif numCerveses>100:
            return True
        else:
            return esCapDeSetmana

        

            
        
        
            
        
    
    

In [38]:
# Test de la funció
prouCerveses(110,True)

# Hauria de retornar: True

True

## Sumatori de parelles

Donada una llista d’enters i  un valor de suma, trobar totes les parelles de nombres a la llista que sumin aquest valor.

Per exemple:
- Llista=[3, 1, 5, 2, 7, 8], Suma=10</li>
- Solució:
   - Parella trobada als índexs 0 i 4,  (3 + 7)</li>
   - Parella trobada als índexs 3 i 5 (2+8)</li>


**Pregunta 1**: Quina solució hauria de donar per als següents valors: 

- Llista:[5, 5, 1, 3, 5], Suma:10
- Llista:[4, 6, 1, 3, 9, 4], Suma:12
- Llista3:[4, 7], Suma:[9]

In [ ]:
llista1=([5,5,1,3,5], 10)
llista2=([4,6,1,3,9,4], 10)
llista3=([4,7], 10)
valorSuma=[10]

In [32]:
def parellesSuma(llista, valorSuma):
    """
    Aquesta funció, donada una llista d’enters i  un valor de suma,
    troba una parella de nombres a la llista que sumin aquest valor.
    :param llista una llista d'enters
    :param valorSuma un enter amb el valor de la suma
    :return: No retorna res, mostra per pantalla el resultat
    """
    for suma in range (1,len (llista)-1):
        for sumacle in range (suma+1, len(llista)):
            if llista[suma]+llista[sumacle]==valorSuma:
    
    
                parellesSuma= llista[suma] , llista[sumacle]
    
    return parellesSuma
    
                
                
                
            

                
        
    

    

    

    

In [33]:
# Test de la funció
parellesSuma([5,5,1,3,5], 10)

# Hauria de retornar: Parella trobada als indexs 0, 4 (3 + 7) 
# i parella trobada als índexs 3 i 5 (2 + 8)

(5, 5)

**Pregunta 2**: Si la llista té 5 posicions, quantes comparacions es fan?

## Divisors

Escriu una funció ``divisors`` que demani un nombre menor que 30 a l'usuari i mostri una llista amb tots els seus divisors.

Exemple:
- Digues-me un nombre? 8  => [1,2,4,8]
- Digues-me un nombre? 5  => [1,5]

In [50]:
def divisors():
    """
    Aquesta funció, pregunta a l'usuari un nombre i retorna
    la llista amb els seus divisors
    :return: llista amb els divisors del nombre introduït per l'usuari
    """
    nombre=31
    while nombre>30:
        n=input ("introduce a number smaller than 30   ")
        nombre=int(n)
    llista= [ ]
    
    for i in range (1,nombre):
        if nombre%i==0:
            llista.append(i)
    
            
       
        
    return llista
            
        
        
    

In [51]:
# Test de la funció
divisors()

# Digues-me un nombre? 28
# Hauria de retornar: [1, 2, 4, 7, 14]

introduce a number smaller than 30   28


[1, 2, 4, 7, 14]

<p style="text-align:right;font-size:0.9em">
&copy;Jordi Vitrià i Mireia Ribera
</p>